packages/modules/vehicles/common/vehicle_utils.py

```diff
@@ -0,0 +1,242 @@
+        return self.logFilter
+
+    def infoLog(self, filter: str, txt: str):
+        if self.logFilter != DEF_LOGFILTER:
+            if filter in self.logFilter or 'A' in self.logFilter:
+                log.info('(' + filter + '): ' + txt)
+        else:
+            log.info(txt + ', filter=' + filter + ', logFilter=' + self.logFilter)
+
+
+# mqtt section
+class mqttUtils:
+
+    def __init__(self, lu: Optional[logUtils] = None):
+        if lu:
+            self.lu = lu
+        else:
+            self.lu = logUtils()
+
+    # write rsp. publish vehicle configuration
+    def write_config(self, topic: str, config: dict):
+        try:
+            self.lu.infoLog('i', "write_config:\n" + dumps(config, ensure_ascii=False, indent=4))
+            Pub().pub(topic, config)
+        except Exception as e:
+            log.exception('Token mqtt write exception ' + str(e))
+
+    # publish a single message to 1883 broker
+    def pub(self, topic: str, payload: str):
+        try:
+            self.lu.infoLog('q', "pub: topic=" + topic + ', payload=' + payload)
+            client = mqtt.Client()
+            client.connect("localhost", 1883, 60)
+            ret, mid = client.publish(topic, payload)
+            if ret != 0:
+                self.lu.infoLog('q', "pub: topic=" + topic +
+                                ', payload=' + payload +
+                                ', ret=' + str(ret) +
+                                ', mid=' + str(mid))
+            client.disconnect()
+        except Exception as e:
+            log.exception("pub mqtt error: %s", e)
+
+    # get mqtt content by topic
+    def mqtt_get_topic(self, topic: str, host: str):
+        try:
+            msg = subscribe.simple(topic, hostname=host)
+        except Exception as e:
+            log.exception("mqtt_get_topic error: %s", e)
+            raise
+        return msg
+
+    # publish CarState via MQTT to 1883 broker
+    def set_CarState(self, ev: int, soc: int, range: float, ts: int):
+        try:
+            msgs = []
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/soc'
+            # self.pub(topic, str(soc))
+            msgs.append({'topic': topic, 'payload': str(soc)})
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/range'
+            # self.pub(topic, str(range))
+            msgs.append({'topic': topic, 'payload': str(range)})
+            topic = 'openWB/set/vehicle/' + str(ev) + '/get/soc_timestamp'
+            _ts = time()
+            # self.pub(topic, str(_ts))
+            msgs.append({'topic': topic, 'payload': str(_ts)})
+            self.lu.infoLog('Q', "set_CarState publish.multiple: msgs=\n" + dumps(msgs))
+            self.lu.infoLog('q', "set_CarState: " + str(soc) + '%/' + str(range) + 'KM@' + str(_ts))
+            publish.multiple(msgs, hostname='localhost', port=1883)
+        except Exception as e:
+            log.exception("set_CarState error: %s", e)
+        return
```

Comment:
Der CarState wird in packages/modules/common/store/_car.py gepublished und sollte auch nur von einer einzigen Stelle im Code gepublished werden.

Reply:
In den meisten SOC-Modulen wird der CarState als return Wert an _car.py zurückgegeben.
Im smarteq Modul werden im WebSocket-API die Daten vom Mercedes-Server gepushed.
Im smarteq-Modul läuft dazu ein Thread, der die Daten asynchron empfängt und in den CarState schreibt.
Kann ich das CarState - Schreiben in _car.py aktiv aufrufen, wenn ja, wie?